### PR TITLE
feat: CU-1ht1xeh custom components in navigation

### PR DESCRIPTION
**Pull request by snqb**
This pull requests adds a piece of functionality for adding custom components to the `Navigation`.
The problem came up with the [Model Overview page card](https://app.clickup.com/t/1ht1xeh), where we need custom list of versions on the Navigation's sidebar:
image

This PR makes it possible to declare custom components in the form of
```children: [
            {
              title: 'Feature List',
              id: '4',
              icon: icons.folder,
            },
            {
              id: '5',
              component: (
                <Box p="10px">
                  <h4>This is a custom component!</h4>
                  <Bar value={0.99} />
                </Box>
              ),
            }
```

I'm not a huge fan of the result, but I really tried to make it as clean as possible, so any feedback on the implementation is welcome.

I've update the Storybook too.
image

### CODE REVIEW

**Comment by giardiv** [comment]
Task linked: [CU-1ht1xeh Model Overview page](https://app.clickup.com/t/1ht1xeh)

File: src/components/navigation/context/getNode.ts [hunk 1]
```diff
@@ -1,17 +1,24 @@
...
-    return tree.reduce(
-      (acc: null | TreeNode, node) => acc || getNode(node, key, value),
+    return tree.reduce<ReturnType>(
+      (acc, node) => acc || getNode(node, key, value),
       null,
     );
   }
 
+  if (isComponentNode(tree)) {
+    return null;
```
Comment (snqb) [inline src/components/navigation/context/getNode.ts]: We return null here, as we don't really need to have those events for the custom components. And we have no good way of doing that anyways too.
Yes, not ideal. But what's ideal in this life?

File: src/components/navigation/types.ts [hunk 2]
```diff
@@ -16,11 +16,15 @@ export interface NavigationItemProps extends Omit<BoxProps, 'css' | 'id'> {
   href?: string;
 }
 
-export interface TreeNode extends NavigationItemProps {
-  id: string;
-  title: string;
```
Comment (snqb) [inline src/components/navigation/types.ts:21]: these were duplicated in the extension of the interface, making it redundant and also kinda misleading

File: src/components/navigation/utils.ts [hunk 3]
```diff
@@ -0,0 +1,5 @@
+import { ComponentNode, TreeNode } from './types';
+
+export default function isComponentNode(tree: TreeNode): tree is ComponentNode {
+  return 'component' in tree;
```
Comment (snqb) [inline src/components/navigation/utils.ts:4]: I'd love to use `tree.hasOwnProperty('component')` as a good practice, but eslint didn't like that. But I guess there's no much difference in such a straightforward case with no prototype chain.

**Review by giardiv — APPROVED** [review]

**Comment by giardiv** [comment]
That works, its not ideal we had to do it like that but the implementation is good

**Comment by snqb** [comment]
I noticed we use `npm`, but without `package-lock.json`
is this intentional?

File: src/components/navigation/context/getNode.ts (hunk 1 above)
Comment (snqb) [reply src/components/navigation/context/getNode.ts]: I was wrong on this, had to redo this

**Comment by guceda** [comment]
> I noticed we use `npm`, but without `package-lock.json`
> is this intentional?

We use `yarn` in Quartz so there is a `yarn-lock.json`.  We recently moved from `npm`,  the documentation you read might not be up to date.

File: src/components/navigation/utils.ts (hunk 3 above)
Comment (guceda) [reply src/components/navigation/utils.ts:4]: Either solutions are fine to me :) 
Would you mind converting this into an expression with an anonymous function to keep consistency with the rest of the code? `const isComponentNode = (tree: TreeNode): tree is ComponentNode => {....}`

**Review by guceda — CHANGES_REQUESTED** [review]
1. I added a comment on `navigation/utils`. 
2. Can you remove the `package-lock.json` file. There is already `yarn`s lock file

**Review by guceda — APPROVED** [review]

**Pull request merged**